jvm/src/main/scala/org/apache/spark/sql/execution/joins/BuildSideRelation.scala

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.joins
+
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+trait BuildSideRelation {
+
+  /**
+   * Deserialized relation from broadcasted value
+   */
+  def deserialized: Iterator[ColumnarBatch]
+
+  /**
+   * Returns a read-only copy of this, to be safely used in current thread.
+   */
+  def asReadOnlyCopy(): BuildSideRelation
```

Comment:
@zzcclp BHJ basically reuses the implementation of SHJ, but the build side is replaced with broadcast relation. You may implement your `BuildSideRelation` to control the behavior of the broadcasted value. For example, make the `asReadOnlyCopy` return valid value for one thread and return empty value for other threads, and in the backend the thread holds valid value will be responsible for building hash table.